### PR TITLE
Optimize queries

### LINKED ISSUE #404: Performance of Icinga DB WEB Query & hints
##  Icinga DB Web QUERY Performance

https://github.com/Icinga/icingadb-web/blob/46c464e3bf0941b0d22f5557f1012e2dd2a692aa/application/controllers/HostController.php#L126-L129

We should use the following filter here instead.

```php
->where([
        'history.host_id = ?' => $this->host->id,
         'history.service_id IS NULL'
]);
```

### Recommended indexes

```sql
ALTER TABLE host_state DROP INDEX idx_host_state_host_id;
ALTER TABLE host_state ADD UNIQUE INDEX idx_host_state_host_id(host_id, severity, last_state_change, soft_state);
ALTER TABLE host_state ADD UNIQUE INDEX idx_host_state_severity(severity, host_id, last_state_change, soft_state);
ALTER TABLE host_state ADD UNIQUE INDEX idx_host_state_soft_state(soft_state, host_id, severity, last_state_change);
ALTER TABLE host_state ADD UNIQUE INDEX idx_host_state_last_state_change(last_state_change, host_id, severity, soft_state);
```
These also applies to the `servicegroup` table.

```sql
ALTER TABLE hostgroup ADD Index idx_hostgroup_display_name(display_name);
ALTER TABLE hostgroup ADD Index idx_hostgroup_name(name);
ALTER TABLE hostgroup ADD Index idx_hostgroup_name_ci(name_ci);
```

`icingadb/services`: Removing [host.display_name](https://github.com/Icinga/icingadb-web/blob/46c464e3bf0941b0d22f5557f1012e2dd2a692aa/application/controllers/ServicesController.php#L63) and [service.display_name](https://github.com/Icinga/icingadb-web/blob/46c464e3bf0941b0d22f5557f1012e2dd2a692aa/application/controllers/ServicesController.php#L67) reduces the result set only to the query limit.

Add new column `host_id` to the `service_state` table.

```sql
ALTER TABLE service_state ADD COLUMN host_id binary(20) NOT NULL COMMENT 'service.host.id',
UPDATE service_state SET host_id= (select service.host_id from service left join host h on h.id = service.host_id where service.id = service_state.service_id);
```

Host services performance with & without the following indexes.

**ORI Query:** 404 ms

**UNIQUE Indexes:** 303 ms

**NON-UNIQUE Indexes:** 391 ms

```sql
ALTER TABLE service_state DROP INDEX idx_service_state_service_id;
ALTER TABLE service_state ADD UNIQUE INDEX idx_service_state_service_id(service_id, host_id, severity, last_state_change, soft_state);
ALTER TABLE service_state ADD UNIQUE INDEX idx_service_state_host_id(host_id, service_id, severity, last_state_change, soft_state);
ALTER TABLE service_state ADD UNIQUE INDEX idx_service_state_severity(severity, host_id, service_id, last_state_change, soft_state);
ALTER TABLE service_state ADD UNIQUE INDEX idx_service_state_soft_state(soft_state, host_id, service_id, severity, last_state_change);
ALTER TABLE service_state ADD UNIQUE INDEX idx_service_state_last_state_change(last_state_change, host_id, service_id, severity, soft_state);
```

https://github.com/Icinga/icingadb-web/blob/46c464e3bf0941b0d22f5557f1012e2dd2a692aa/application/controllers/HostController.php#L197

We can then use the newly introduced `host_id` column in `service_state` as follows here. This way the query would be executed much faster.

```php
$filter = ['service_state.host_id = ?' => $this->host->id];
if ($this->getParam('sort') === 'service.display_name') {
     $filter = ['service.host_id = ?' => $this->host->id];
}
```

Adding `host_id` to `service_state` causes an ambiguity error in where clause in host detail view. Therefore, `host_id`
must be prefixed with service [here](https://github.com/Icinga/icingadb-web/blob/46c464e3bf0941b0d22f5557f1012e2dd2a692aa/application/controllers/HostController.php#L64).

### Services view

```sql
ALTER TABLE service ADD INDEX idx_service_is_volatile(is_volatile, perfdata_enabled) COMMENT 'Service list filtered by is_volatile';
ALTER TABLE service ADD INDEX idx_service_perfdata_enabled(perfdata_enabled, is_volatile) COMMENT 'Service list filtered by perfdata_enabled';
```

**?????**: Adding the following indexes to the service & host tables can reduce the result set by half, but not sure if it's even worth it. After all, we need to add an index to all host and service columns, which you can filter by in the filter editor.

````sql
ALTER TABLE service ADD INDEX idx_service_is_volatile(is_volatile, perfdata_enabled) COMMENT 'Service list filtered by is_volatile ordered by columns from service_state';
ALTER TABLE service ADD INDEX idx_service_perfdata_enabled(perfdata_enabled, is_volatile) COMMENT 'Service list filtered by perfdata_enabled ordered by columns from service_state';
ALTER TABLE service ADD INDEX idx_service_active_check_enabled(active_checks_enabled, passive_checks_enabled, notifications_enabled, event_handler_enabled, flapping_enabled) COMMENT 'Service list filtered by active_checks_enabled ordered by columns from service_state';
ALTER TABLE service ADD INDEX idx_service_passive_check_enabled(passive_checks_enabled, active_checks_enabled, notifications_enabled, event_handler_enabled, flapping_enabled) COMMENT 'Service list filtered by passive_checks_enabled ordered by columns from service_state';
ALTER TABLE service ADD INDEX idx_service_notifications_enabled(notifications_enabled, active_checks_enabled, passive_checks_enabled, event_handler_enabled, flapping_enabled) COMMENT 'Service list filtered by notifications_enabled ordered by columns from service_state';
ALTER TABLE service ADD INDEX idx_service_event_handler_enabled(event_handler_enabled, notifications_enabled, active_checks_enabled, passive_checks_enabled, flapping_enabled) COMMENT 'Service list filtered by event_handler_enabled ordered by columns from service_state';
ALTER TABLE service ADD INDEX idx_service_flapping_enabled(flapping_enabled, event_handler_enabled, notifications_enabled, active_checks_enabled, passive_checks_enabled) COMMENT 'Service list filtered by flapping_enabled ordered by columns from service_state';
````

We should add the following suggestion to [this line](https://github.com/Icinga/icingadb-web/blob/46c464e3bf0941b0d22f5557f1012e2dd2a692aa/application/controllers/ServicesController.php#L57), to allow it to use the `host_id` index already when joining to the `service_state` table before sorting by `host.display_name`.

```php
if ($this->getParam('sort') === 'host.display_name') {
      $services->getSelectBase()->where('service_state.host_id = service_host.id');
}
```

### Downtimes

```sql
... FROM downtime
         LEFT JOIN host downtime_host ON downtime_host.id = downtime.host_id
         LEFT JOIN host_state downtime_host_state ON downtime_host_state.host_id = downtime_host.id
         LEFT JOIN service downtime_service ON downtime_service.id = downtime.service_id
         LEFT JOIN host downtime_service_host ON downtime_service_host.id = downtime_service.host_id
         LEFT JOIN host_state downtime_service_host_state
                   ON downtime_service_host_state.host_id = downtime_service_host.id
         LEFT JOIN service_state downtime_service_state ON downtime_service_state.service_id = downtime_service.id
```
Is there any reason here why we are joining the `host` and `host_state` table twice? These are also joined in the same way within the Comments view. cc @nilmerg 

```sql
ALTER TABLE downtime DROP INDEX idx_downtime_is_in_effect;
ALTER TABLE downtime ADD INDEX idx_downtime_is_in_effect(is_in_effect, start_time, end_time, entry_time, scheduled_start_time, scheduled_end_time, author);
ALTER TABLE downtime ADD  INDEX idx_downtime_start_time(start_time, is_in_effect, end_time, entry_time, scheduled_start_time, scheduled_end_time, author);
ALTER TABLE downtime ADD  INDEX idx_downtime_end_time(end_time, is_in_effect, start_time, entry_time, scheduled_start_time, scheduled_end_time, author);
ALTER TABLE downtime ADD  INDEX idx_downtime_entry_time(entry_time, is_in_effect, start_time, end_time, scheduled_start_time, scheduled_end_time, author);
ALTER TABLE downtime ADD  INDEX idx_downtime_scheduled_start_time(scheduled_start_time, is_in_effect, start_time, end_time, entry_time, scheduled_end_time, author);
ALTER TABLE downtime ADD  INDEX idx_downtime_scheduled_end_time(scheduled_end_time, is_in_effect, start_time, end_time, entry_time, scheduled_start_time, author);
ALTER TABLE downtime ADD  INDEX idx_downtime_author(author, is_in_effect, start_time, end_time, entry_time, scheduled_start_time, scheduled_end_time);
```

We should remove the sort by direction by default from [here](https://github.com/Icinga/icingadb-web/blob/46c464e3bf0941b0d22f5557f1012e2dd2a692aa/application/controllers/DowntimesController.php#L46) and we should add a new column to the `downtime` table in order to avoid such a query when sorting downtimes by duration.

```sql
... ORDER BY CASE WHEN is_flexible = 'y' THEN flexible_duration ELSE scheduled_end_time - scheduled_start_time END desc;
```

```sql
ALTER TABLE downtime ADD COLUMN duration bigint unsigned NOT NULL COMMENT 'Duration of the downtime: When the downtime is flexible, this is the same as flexible_duration otherwise scheduled_end_time - scheduled_start_time';
```

### Comments

We could add these to the comment table but have not managed to get mysql to use one of these indexes. ?????

```sql
ALTER TABLE comment DROP INDEX idx_comment_entry_time;
ALTER TABLE comment ADD INDEX idx_comment_entry_time(entry_time, expire_time, author) COMMENT 'Comment list fileted/ordered by entry_time';
ALTER TABLE comment ADD  INDEX idx_comment_expire_time(expire_time, entry_time, author) COMMENT 'Comment list fileted/ordered by expire_time';
ALTER TABLE comment ADD  INDEX idx_comment_author(author, entry_time, expire_time) COMMENT 'Comment list fileted/ordered by author';
```

### History

```sql
ALTER TABLE history DROP INDEX idx_history_event_time;
ALTER TABLE history ADD INDEX idx_history_event_time (event_time, service_id, host_id) COMMENT 'History filtered/ordered by event_time';
ALTER TABLE history ADD INDEX idx_history_host_id (host_id, service_id, event_time) COMMENT 'History filtered/ordered by host_id';
ALTER TABLE history ADD INDEX idx_history_service_id (service_id, host_id, event_time) COMMENT 'History list filtered/ordered by service_id';
```

Why do we join the host table with `inner join` in the history view? Since inner joins are executed before where clause,
all hosts are selected first before the filter is applied and this is IMO an overkill. I know probably the reason why
hosts always have to be there, and then we have to just inner join on the host table but that slows down the query by 8 times.
Besides, why do we filter by default for `event_time`? That is also completely superfluous I think.

Compares: With 5K hosts, 15K+ services and 20K+ histories

| LEFT JOIN ON host table | INNER JOIN ON host table | W/ filtering for event_time |
| :---:                   |    :----:                |          :---               |
| 430 ms                  | 4s 340ms                 | no                          |
| 570ms                   | 5s 644ms                 | yes                         |

### PR DESCRIPTION
**Pull request by Yonas-net**
resolves #404

### CODE REVIEW

**Comment by Yonas-net** [comment]
**Blocked by** https://github.com/Icinga/icinga2/issues/9049

**Comment by Yonas-net** [comment]
Needs just to be tested properly after these PRs are merged.
https://github.com/Icinga/icinga2/pull/9061
https://github.com/Icinga/icingadb/pull/403
https://github.com/Icinga/icingadb/pull/397

**Comment by lippserd** [comment]
Despite Icinga/icingadb#397 not being merged yet, this one is ready to review.

File: application/controllers/HistoryController.php
Comment (nilmerg) [inline application/controllers/HistoryController.php:85]: What if there's a host (`history_host.id IS NOT NULL`) for the service history entry, but no service? (`history_service.id IS NULL`)

File: application/controllers/HostsController.php
Comment (nilmerg) [inline application/controllers/HostsController.php:41]: Why not directly in the model?

File: application/controllers/ServicesController.php
Comment (nilmerg) [inline application/controllers/ServicesController.php:50]: Also here, why not in the model?

File: application/controllers/HostsController.php
Comment (lippserd) [reply application/controllers/HostsController.php:41]: I can't give you a 100 percent answer here, but that can slow down other queries.

**Review by nilmerg — APPROVED** [review]

**Pull request merged**
